src/components/app/BottomBox.js

```diff
@@ -107,47 +180,112 @@ class BottomBoxImpl extends React.PureComponent<Props> {
+            <div className="bottom-box-bar">
+              <h3 className="bottom-box-title">{path ?? '(no source file)'}</h3>
+              {assemblyViewIsOpen ? null : trailingHeaderButtons}
+            </div>
+            <div className="bottom-sourceview-wrapper">
+              {sourceViewFile !== null ? (
+                <SourceView
+                  disableOverscan={disableOverscan}
+                  timings={globalLineTimings}
+                  sourceCode={sourceCode}
+                  filePath={path}
+                  scrollToHotSpotGeneration={sourceViewScrollGeneration}
+                  hotSpotTimings={selectedCallNodeLineTimings}
+                  ref={this._sourceView}
+                />
+              ) : null}
+              {sourceViewCode !== undefined &&
+              sourceViewCode.type === 'LOADING' ? (
+                <CodeLoadingOverlay source={sourceViewCode.source} />
+              ) : null}
+              {sourceViewCode !== undefined &&
+              sourceViewCode.type === 'ERROR' ? (
+                <SourceCodeErrorOverlay errors={sourceViewCode.errors} />
+              ) : null}
+            </div>
+          </div>
+
+          {assemblyViewIsOpen ? (
+            <div className="bottom-box-pane">
+              <div className="bottom-box-bar">
+                <h3 className="bottom-box-title">
+                  {assemblyViewNativeSymbol !== null
+                    ? assemblyViewNativeSymbol.name
+                    : '(no native symbol)'}
+                </h3>
+                {trailingHeaderButtons}
+              </div>
+              <div className="bottom-assemblyview-wrapper">
+                {assemblyViewNativeSymbol !== null ? (
+                  <AssemblyView
+                    disableOverscan={disableOverscan}
+                    timings={globalAddressTimings}
+                    assemblyCode={assemblyCode}
+                    nativeSymbol={assemblyViewNativeSymbol}
+                    scrollToHotSpotGeneration={assemblyViewScrollGeneration}
+                    hotSpotTimings={selectedCallNodeAddressTimings}
+                    ref={this._assemblyView}
+                  />
+                ) : null}
+                {assemblyViewCode !== undefined &&
+                assemblyViewCode.type === 'LOADING' ? (
+                  <CodeLoadingOverlay source={assemblyViewCode.source} />
+                ) : null}
+                {assemblyViewCode !== undefined &&
+                assemblyViewCode.type === 'ERROR' ? (
+                  <AssemblyCodeErrorOverlay
+                    errors={convertErrors(assemblyViewCode.errors)}
+                  />
+                ) : null}
+              </div>
+            </div>
           ) : null}
-        </div>
+        </SplitterLayout>
       </div>
     );
   }
 }
 
+function convertErrors(errors: ApiQueryError[]): SourceCodeLoadingError[] {
+  // Copy the array so that the types work out.
+  return errors.map((e) => e);
```

Comment:
Hmm this looks a bit unnecessary. I wonder if we can let flow understand this without copying. I wanted to check it locally but looks like I need to fix my flow setup first :')

Reply:
I think it's necessary unless there's a way to tell Flow that we will not mutate the result.

If we don't copy the array here, I can see why Flow doesn't like it.

```js
const apiErrors = [{ type: 'BROWSER_API_ERROR', apiErrorMessage: '' }];
const sourceCodeLoadingErrors = convertErrors(apiErrors);
sourceCodeLoadingErrors.push({ type: 'NO_KNOWN_CORS_URL' });
// Now apiErrors has an object in it which is not allowed by the ApiQueryError type!
```

Reply:
@mstange This seems to work for me with a `$ReadOnlyArray`:

```diff
diff --git a/src/components/app/BottomBox.js b/src/components/app/BottomBox.js
index 348ecd048..27b882283 100644
--- a/src/components/app/BottomBox.js
+++ b/src/components/app/BottomBox.js
@@ -34,18 +34,16 @@ import explicitConnect from 'firefox-profiler/utils/connect';
 
 import type { ConnectedProps } from 'firefox-profiler/utils/connect';
 import type {
   LineTimings,
   AddressTimings,
   SourceCodeStatus,
   AssemblyCodeStatus,
   NativeSymbolInfo,
-  SourceCodeLoadingError,
-  ApiQueryError,
 } from 'firefox-profiler/types';
 import type { CodeErrorOverlayProps } from './CodeErrorOverlay';
 
 import { Localized } from '@fluent/react';
 
 import './BottomBox.css';
 
 type StateProps = {|
@@ -257,34 +255,27 @@ class BottomBoxImpl extends React.PureComponent<Props> {
                   />
                 ) : null}
                 {assemblyViewCode !== undefined &&
                 assemblyViewCode.type === 'LOADING' ? (
                   <CodeLoadingOverlay source={assemblyViewCode.source} />
                 ) : null}
                 {assemblyViewCode !== undefined &&
                 assemblyViewCode.type === 'ERROR' ? (
-                  <AssemblyCodeErrorOverlay
-                    errors={convertErrors(assemblyViewCode.errors)}
-                  />
+                  <AssemblyCodeErrorOverlay errors={assemblyViewCode.errors} />
                 ) : null}
               </div>
             </div>
           ) : null}
         </SplitterLayout>
       </div>
     );
   }
 }
 
-function convertErrors(errors: ApiQueryError[]): SourceCodeLoadingError[] {
-  // Copy the array so that the types work out.
-  return errors.map((e) => e);
-}
-
 export const BottomBox = explicitConnect<{||}, StateProps, DispatchProps>({
   mapStateToProps: (state) => ({
     sourceViewFile: getSourceViewFile(state),
     sourceViewCode: getSourceViewCode(state),
     globalLineTimings: selectedThreadSelectors.getSourceViewLineTimings(state),
     selectedCallNodeLineTimings:
       selectedNodeSelectors.getSourceViewLineTimings(state),
     sourceViewScrollGeneration: getSourceViewScrollGeneration(state),
diff --git a/src/components/app/CodeErrorOverlay.js b/src/components/app/CodeErrorOverlay.js
index f833fd19c..f647c1db9 100644
--- a/src/components/app/CodeErrorOverlay.js
+++ b/src/components/app/CodeErrorOverlay.js
@@ -5,17 +5,17 @@
 
 import React from 'react';
 
 import { assertExhaustiveCheck } from 'firefox-profiler/utils/flow';
 import type { SourceCodeLoadingError } from 'firefox-profiler/types';
 import { Localized } from '@fluent/react';
 
 export type CodeErrorOverlayProps = {|
-  errors: SourceCodeLoadingError[],
+  errors: $ReadOnlyArray<SourceCodeLoadingError>,
 |};
 
 export function CodeErrorOverlay({ errors }: CodeErrorOverlayProps) {
   return (
     <ul className="codeErrorOverlay">
       {errors.map((error, key) => {
         switch (error.type) {
           case 'NO_KNOWN_CORS_URL': {
```

Am I missing something?

Reply:
Nice! I wasn't aware of `$ReadOnlyArray`. I had briefly tried `$ReadOnly<SourceCodeLoadingError[]>` but I've never really worked with these special types and didn't check the documentation.

Do you want to submit this patch as a PR or would you like me to do it?

Reply:
I cleaned up my local copy already oops :) I'm happy if you can do it and double check that I didn't forget anything, and I can review!